### PR TITLE
Mvp human vs human

### PR DESCRIPTION
**Pull request by Aoxian**
I have a working mvp-human-vs-human TTT in elixir.  I’m not in love with my implementation, but it does work.  Really looking forward to the feedback from you all.

### CODE REVIEW

File: lib/game.ex [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {board, :game_over}
+  end
+
+  def process_move(board, move, mark) do
+    {:ok, board} = Board.place_mark(board, move, mark)
+    {board, :playing}
+  end
+
+  def swap_mark("X"), do: "O"
+  def swap_mark("O"), do: "X"
```
Comment (td-extreme) [inline lib/game.ex]: I'm not a big fan of this functions name. When I see `swap` in code it makes me think of two object switching places in memory. I would think a function with a name like this might do something like take a map of config vars, and return a map where the makers of each player have been swapped. Change the configs so that now player one will be displayed as `O` and player two as `X`. 
What you function is doing is taking a player and giving me the other player back. You might be using it to toggle or switch players in your recursive loop, but this function doesn't really swap anything.
Something like `other_player` or `opponent`, `next_player` would better describe what this function does.

File: test/game_test.exs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      :display_board_called
+    end
+
+    def display_message(_message) do
+      :display_message_called
+    end
+  end
+
+  test "display_board() calls the UI to display the board" do
+    assert Game.display_board(Board.empty(), MockConsole) == :display_board_called
```
Comment (td-extreme) [inline test/game_test.exs]: I assume you're trying to mock this partial from our talk the other day about wrapping and mocking IO. What you have here isn't really testing anything.  Ideally you would have a mock and test that when you call `game_loop` that it calls the method in your mock. For this PR I'm going with you not testing `IO`.

File: lib/game.ex (hunk 1 above)
Comment (Aoxian) [reply lib/game.ex]: Renamed to next_player.

File: test/game_test.exs (hunk 2 above)
Comment (Aoxian) [reply test/game_test.exs]: I remove this `mock` and the associated tests.

File: lib/board.ex [hunk 3]
```diff
@@ -15,11 +15,30 @@ defmodule Board do
     }
   end
 
+  def full?(board = %Board{}) do
+    board
+    |> Map.from_struct()
+    |> Map.values()
+    |> Enum.filter(fn x -> String.match?(x, ~r/[1..9]/) end) == []
```
Comment (7hoenix) [inline lib/board.ex]: What is that x in the anonymous function?

Maybe we can just extract it to a named function?

File: lib/console_printer.ex [hunk 4]
```diff
@@ -12,4 +12,10 @@ defmodule ConsolePrinter do
     |> format_message
     |> IO.puts()
   end
+
+  def get_player_move(message) do
+    [move | _] = IO.gets(message) |> String.split("\n")
```
Comment (7hoenix) [inline lib/console_printer.ex]: Sounds like this is out of scope for this MR.

But if you create your own version of IO... and inject it.

Then you will be able to test it.

In the actual production code you would use it like:

```
ConsoleUI.get_player_move(IO.gets, message)
```

And in your test code you could do something like:

```
ConsoleUI.get_player_move(my_mock_function, message)
```

File: test/board_test.exs [hunk 5]
```diff
@@ -32,4 +32,72 @@ defmodule BoardTest do
...
+    assert Board.decode_key("9") == {:ok, :nine}
+  end
+
+  test "Decode a q to :quit" do
+    assert Board.decode_key("q") == {:ok, :quit}
+  end
+
+  test "Invalid keys sent to decode_key return :invalid_key message" do
+    assert Board.decode_key("A") == {:error, :invalid_key}
+  end
```
Comment (7hoenix) [inline test/board_test.exs]: All of these tests are the same.

Makes me feel like a property based test would work well here (have not used them yet in Elixir though).

These tests also feel kind of bad to me though...

Like what if we want to make a vim version of the game (where I don't need to leave the homerow).

All of these tests would need to be rewritten.

Maybe just two tests could give us exactly the same thing:

```
test "good keys" do
  val good_keys = [{:hi, "ok"}]
  assert Board.decode_key(good_keys, :hi) === {:ok,"ok"}
end

test "bad keys" do
 assert Board.decode_key([], :hi) === {:error, :invalid_key} 
end
```

Notice how injecting the configurable part is the trick that enables this.

File: lib/board.ex (hunk 3 above)
Comment (Aoxian) [reply lib/board.ex]: I created a named function for this called: Board.initial_value?(value)

File: lib/console_printer.ex (hunk 4 above)
Comment (Aoxian) [reply lib/console_printer.ex]: It is out of scope (for now), but I plan on figuring the injection stuff out and start using it.

File: test/board_test.exs (hunk 5 above)
Comment (Aoxian) [reply test/board_test.exs]: I've removed the duplicate tests as suggested.

I do not understand your vim example though, as I am not familiar with vim games or the concept of not leaving the homerow.

File: lib/board.ex [hunk 6]
```diff
@@ -15,11 +15,34 @@ defmodule Board do
     }
   end
 
+  def full?(board = %Board{}) do
+    board
+    |> Map.from_struct()
+    |> Map.values()
+    |> Enum.filter(fn value -> initial_value?(value) end) == []
```
Comment (7hoenix) [inline lib/board.ex]: You should be able to just write -> `Enum.filter(initial_value?)`

File: lib/board.ex [hunk 7]
```diff
@@ -15,11 +15,34 @@ defmodule Board do
...
+  end
+
+  def initial_value?(value) do
+    String.match?(value, ~r/[1..9]/)
```
Comment (7hoenix) [inline lib/board.ex]: Maybe -> `def empty_square?(square) do`?

File: lib/game_ui.ex [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  require Board
+
+  def display_board(state) do
+    state.ui.display_board(state.board)
+    state
+  end
+
+  def display_message(state, message) do
+    state.ui.display_message(message)
+    state
```
Comment (7hoenix) [inline lib/game_ui.ex]: Are we doing this so that we are able to execute the side effect, but also stay inside the pipeline?

File: lib/board.ex (hunk 6 above)
Comment (Aoxian) [reply lib/board.ex]: When attempting this, I end up with compilation errors.
`undefined function initial_value?/0`

If I follow it up with a `Enum.filter(initial_value?(value))` then it still complains, all the way until the function call evolves to that anonymous function.  I think it has something to do with how the variable is passed into it via that `fn var ->' portion.

File: lib/board.ex (hunk 7 above)
Comment (Aoxian) [reply lib/board.ex]: I renamed the function.

File: lib/game_ui.ex (hunk 8 above)
Comment (Aoxian) [reply lib/game_ui.ex]: If by 'this' you mean returning the state at the end of the function, then yes.

File: lib/board.ex (hunk 6 above)
Comment (7hoenix) [reply lib/board.ex]: Ah. In elixir it needs the arity as well. Try -> `Enum.filter(initial_value?/1)`

**Comment by 7hoenix** [comment]
I'm good with the changes as they stand now.

Overall seems like a good start.

File: lib/board.ex (hunk 6 above)
Comment (Aoxian) [reply lib/board.ex]: I tried the second suggetion, but the elixir formatter doesn't allow for that either.  I'm with you, it seems like there should be some syntactical sugar to allow for it, but its not working the way we would like.

Comment (7hoenix) [reply lib/board.ex]: @nystromb  Please halp
